dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -344,23 +347,58 @@ def read_metadata(
         # Create `parts`
         # This is a list of row-group-descriptor dicts, or file-paths
         # if we have a list of files and gather_statistics=False
+        base_path = (base_path or "") + fs.sep
         if not parts:
+            # Populate `partsin` with dataset row-groups
             partsin = pf.row_groups
-            if fast_metadata:
-                pf = (paths, gather_statistics)
+            partitions = pf.info.get("partitions", None)
+            if partitions and not fast_metadata:
+                # We have partitions, and do not have
+                # a "_metadata" file for the worker to read.
+                # Therefore, we need to pass the pf object in
+                # the task graph
+                pf_deps = pf
+            else:
+                # We don't need to pass a pf object in the task graph.
+                # Instead, we can try to pass the path for each part.
+                pf_deps = "tuple"
         else:
+            # `parts` is just a list of path names.
+            # No tricky logic necessary
+            partitions = None
+            pf_deps = None
             partsin = parts
-            pf = None
         parts = []
+        i_path = 0
+        path_last = None
         for i, piece in enumerate(partsin):
-            if pf and not fast_metadata:
-                for col in piece.columns:
-                    col.meta_data.statistics = None
-                    col.meta_data.encoding_stats = None
-            piece_item = i if pf else piece
+            if partitions and fast_metadata:
```

Comment:
The three conditions here are a bit hard to parse. Of course, the situation **is** complicated. Perhaps up front, we should enumerate the cases and label the branches as required:
- a single file
- a directory of files with a _metadata
- a directory of files without _metadata
  - stats requested
  - stats unnecessary